src/can.rs

```diff
@@ -0,0 +1,498 @@
+
+        // Time seg 2
+        // tBS2 = tq x (TS2[2:0] + 1)
+        let ts2 = 3;
+
+        // Time seg 1
+        // tBS1 = tq x (TS1[3:0] + 1)
+        let ts1 = 10;
+
+        can.btr.modify(|_, w| unsafe {
+            w.brp()
+                .bits(brp)
+                .sjw()
+                .bits(sjw)
+                .ts1()
+                .bits(ts1)
+                .ts2()
+                .bits(ts2)
+            //.lbkm()
+            //.set_bit()
+        });
+
+        // Leave initialization mode by clearing INRQ and switch to normal mode
+        can.mcr.modify(|_, w| w.inrq().clear_bit());
+        // Wait for INAK again to confirm the mode switch
+        while !can.msr.read().inak().bit_is_clear() {}
+
+        Can {
+            can,
+            _rx: rx,
+            _tx: tx,
+        }
+    }
+
+    pub fn listen(&mut self, event: Event) {
+        self.can.ier.modify(|_, w| match event {
+            Event::Fifo0Fmp => w.fmpie0().set_bit(),
+            Event::Fifo0Full => w.ffie0().set_bit(),
+            Event::Fifo0Ovr => w.fovie0().set_bit(),
+            Event::Txe => w.tmeie().set_bit(),
+            _ => unimplemented!(),
+        });
+    }
+
+    pub fn split(self) -> (CanTransmitter, CanFifo, CanFifo) {
+        let fifo0 = CanFifo { idx: 0 };
+        let fifo1 = CanFifo { idx: 1 };
+
+        let transmitter = CanTransmitter {
+            _can: self.can,
+            _rx: self._rx,
+            _tx: self._tx,
+        };
+
+        return (transmitter, fifo0, fifo1);
+    }
+
+    pub fn release(self) -> stm32::CAN {
+        self.can
+    }
+}
+
+impl embedded_hal_can::Interface for CanTransmitter {
+    type Id = CanId;
+    type Frame = CanFrame;
+    type Error = ();
+    type Filter = CanFilter;
+}
+
+impl embedded_hal_can::Interface for CanFifo {
+    type Id = CanId;
+    type Frame = CanFrame;
+    type Error = ();
+    type Filter = CanFilter;
+}
+
+impl embedded_hal_can::Transmitter for CanTransmitter {
+    fn transmit(
+        &mut self,
+        frame: &Self::Frame,
+    ) -> Result<Option<Self::Frame>, nb::Error<Self::Error>> {
+        let can = can_register();
+
+        for tx_idx in 0..3 {
+            let free = match tx_idx {
+                0 => can.tsr.read().tme0().bit_is_set(),
+                1 => can.tsr.read().tme1().bit_is_set(),
+                2 => can.tsr.read().tme2().bit_is_set(),
+                _ => unreachable!(),
+            };
+
+            if !free {
+                continue;
+            }
+
+            let tx = &can.tx[tx_idx];
+
+            match frame.id {
+                CanId::BaseId(id) => tx.tir.modify(|_, w| unsafe { w.stid().bits(id) }),
+                CanId::ExtendedId(id) => tx.tir.modify(|_, w| unsafe {
+                    let hb = ((id & EXID_MASK) >> 18) as u16;
+                    w.exid().bits(id).stid().bits(hb).ide().set_bit()
+                }),
+            }
+
+            if let Some(_) = frame.data() {
+                unsafe {
```

Comment:
A comment on this `unsafe` instance would be great. And couldn't this `unsafe` scope be limited insid3 the `math j { ... }`?

Reply:
I'll add a note, there are actually 2 unsafe calls here which is the reason for the outer scope (the tdtr call below also calls the unsafe `bits`) but these can be split. The note here will probably be somewhat silly, I'd have to do some more digging but I wonder if we could get svd-to-rust to generate safe calls for these registers, as the only actual way to update them is via a full register write